contrib/uriparser/utils/utils_uriparser.cc

```diff
@@ -0,0 +1,286 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "contrib/uriparser/utils/utils_uriparser.h"
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#include "absl/cleanup/cleanup.h"
+
+Uriparser::~Uriparser() {
+  if (GetStatus().ok()) {
+    api_.uriFreeUriMembersA(uri_.PtrBefore()).IgnoreError();
+  }
+};
+
+absl::Status Uriparser::GetStatus() { return status_; }
+
+absl::Status Uriparser::ParseUri() {
+  SAPI_RETURN_IF_ERROR(sandbox_->Allocate(&uri_, true));
+
+  sapi::v::Struct<UriParserStateA> state;
+  state.mutable_data()->uri = reinterpret_cast<UriUriA*>(uri_.GetRemote());
+  SAPI_ASSIGN_OR_RETURN(
+      int ret, api_.uriParseUriA(state.PtrBefore(), c_uri_.PtrBefore()));
+  if (ret != 0) {
+    return absl::UnavailableError("Unable to parse uri");
+  }
+
+  SAPI_RETURN_IF_ERROR(sandbox_->TransferFromSandboxee(&uri_));
+
+  return absl::OkStatus();
+}
+
+absl::StatusOr<std::string> Uriparser::FetchUriText(UriTextRangeA* ptr) {
+  if (ptr == nullptr) {
+    return "";
+  }
+  if (ptr->first == nullptr) {
+    return "";
+  }
+
+  size_t size = ptr->afterLast - ptr->first;
+
+  // Sometimes uriparser dosen't allocate new memory
+  // and sometimes it does.
+  SAPI_ASSIGN_OR_RETURN(
+      std::string uri,
+      sandbox_->GetCString(sapi::v::RemotePtr(const_cast<char*>(ptr->first))));
+
+  return uri.substr(0, size);
+}
+
+absl::StatusOr<std::string> Uriparser::GetScheme() {
+  return FetchUriText(&uri_.mutable_data()->scheme);
+}
+
+absl::StatusOr<std::string> Uriparser::GetUserInfo() {
+  return FetchUriText(&uri_.mutable_data()->userInfo);
+}
+
+absl::StatusOr<std::string> Uriparser::GetHostText() {
+  return FetchUriText(&uri_.mutable_data()->hostText);
+}
+
+absl::StatusOr<std::string> Uriparser::GetHostIP() {
+  char ipstr[INET6_ADDRSTRLEN] = "";
+
+  UriHostDataA* data = &uri_.mutable_data()->hostData;
+
+  if (uri_.mutable_data()->hostData.ip4) {
+    sapi::v::Struct<UriIp4> ip4;
+    ip4.SetRemote(uri_.mutable_data()->hostData.ip4);
+    SAPI_RETURN_IF_ERROR(sandbox_->TransferFromSandboxee(&ip4));
+    inet_ntop(AF_INET, ip4.mutable_data()->data, ipstr, sizeof(ipstr));
+  } else if (uri_.mutable_data()->hostData.ip6) {
```

Comment:
We could get rid of the else if by returning in the if block above and move the else if block outside the conditional.

Reply:
I just realized, that this might be to ensure the case where neither of the ip4 or ip6 fields are set. In that case ignore my previous comment.